platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/maps/Projection.java

```diff
@@ -97,54 +100,119 @@ public VisibleRegion getVisibleRegion() {
+
     LatLng topLeft = fromScreenLocation(new PointF(left, top));
     LatLng topRight = fromScreenLocation(new PointF(right, top));
     LatLng bottomRight = fromScreenLocation(new PointF(right, bottom));
     LatLng bottomLeft = fromScreenLocation(new PointF(left, bottom));
 
-    // Map can be rotated, find correct LatLngBounds that encompasses the visible region (that might be rotated)
-    List<LatLng> boundsPoints = new ArrayList<>();
-    boundsPoints.add(topLeft);
-    boundsPoints.add(topRight);
-    boundsPoints.add(bottomRight);
-    boundsPoints.add(bottomLeft);
-
-    // order so that two most northern point are put first
-    while ((boundsPoints.get(0).getLatitude() < boundsPoints.get(3).getLatitude())
-      || (boundsPoints.get(1).getLatitude() < boundsPoints.get(2).getLatitude())) {
-      LatLng first = boundsPoints.remove(0);
-      boundsPoints.add(first);
+    latLngs.add(topRight);
+    latLngs.add(bottomRight);
+    latLngs.add(bottomLeft);
+    latLngs.add(topLeft);
+
+    double maxEastLonSpan = 0;
+    double maxWestLonSpan = 0;
+
+    double east = 0;
+    double west = 0;
+    double north = GeometryConstants.MIN_LATITUDE;
+    double south = GeometryConstants.MAX_LATITUDE;
+
+    for (LatLng latLng : latLngs) {
+      Point latLngPoint = Point.fromLngLat(latLng.getLongitude(), latLng.getLatitude());
+      double bearing = bearing(centerPoint, latLngPoint);
+
+      if (bearing >= 0) {
+        double span = getLongitudeSpan(latLng.getLongitude(), center.getLongitude());
+        if (span > maxEastLonSpan) {
+          maxEastLonSpan = span;
+          east = latLng.getLongitude();
+        }
+      } else {
+        double span = getLongitudeSpan(center.getLongitude(), latLng.getLongitude());
+        if (span > maxWestLonSpan) {
+          maxWestLonSpan = span;
+          west = latLng.getLongitude();
+        }
+      }
+
+      if (north < latLng.getLatitude()) {
+        north = latLng.getLatitude();
+      }
+      if (south > latLng.getLatitude()) {
+        south = latLng.getLatitude();
+      }
     }
 
-    double north = boundsPoints.get(0).getLatitude();
-    if (north < boundsPoints.get(1).getLatitude()) {
-      north = boundsPoints.get(1).getLatitude();
-    }
+    return new VisibleRegion(topLeft, topRight, bottomLeft, bottomRight,
+      LatLngBounds.from(north, east, south, west));
 
-    double south = boundsPoints.get(2).getLatitude();
-    if (south > boundsPoints.get(3).getLatitude()) {
-      south = boundsPoints.get(3).getLatitude();
-    }
+  }
+
+
+  /**
+   * Takes two {@link Point}s and finds the geographic bearing between them.
+   *
+   * @param point1 first point used for calculating the bearing
+   * @param point2 second point used for calculating the bearing
+   * @return bearing in decimal degrees
+   * @see <a href="http://turfjs.org/docs/#bearing">Turf Bearing documentation</a>
+   */
+  static double bearing(@NonNull Point point1, @NonNull Point point2) {
```

Comment:
Could we change this method's declaration to work with `LatLng`s? We would skip unnecessary creation of the `Point` object before invoking this method.

Reply:
good point! Thank you.